flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java

```diff
@@ -0,0 +1,207 @@
+
+/**
+ * This state factory wraps state objects, produced by backends, with TTL logic.
+ */
+public class TtlStateFactory {
+	public static <N, SV, S extends State, IS extends S> IS createStateAndWrapWithTtlIfEnabled(
+		TypeSerializer<N> namespaceSerializer,
+		StateDescriptor<S, SV> stateDesc,
+		KeyedStateFactory originalStateFactory,
+		TtlConfig ttlConfig,
+		TtlTimeProvider timeProvider) throws Exception {
+		return ttlConfig.getTtlUpdateType() == TtlUpdateType.Disabled ?
+			originalStateFactory.createState(namespaceSerializer, stateDesc) :
+			new TtlStateFactory(originalStateFactory, ttlConfig, timeProvider)
+				.createState(namespaceSerializer, stateDesc);
+	}
+
+	private final Map<Class<? extends StateDescriptor>, StateFactory> stateFactories;
+
+	private final KeyedStateFactory originalStateFactory;
+	private final TtlConfig ttlConfig;
+	private final TtlTimeProvider timeProvider;
+
+	private TtlStateFactory(KeyedStateFactory originalStateFactory, TtlConfig ttlConfig, TtlTimeProvider timeProvider) {
+		this.originalStateFactory = originalStateFactory;
+		this.ttlConfig = ttlConfig;
+		this.timeProvider = timeProvider;
+		this.stateFactories = createStateFactories();
+	}
+
+	private Map<Class<? extends StateDescriptor>, StateFactory> createStateFactories() {
+		return Stream.of(
+			Tuple2.of(ValueStateDescriptor.class, (StateFactory) this::createValueState),
+			Tuple2.of(ListStateDescriptor.class, (StateFactory) this::createListState),
+			Tuple2.of(MapStateDescriptor.class, (StateFactory) this::createMapState),
+			Tuple2.of(ReducingStateDescriptor.class, (StateFactory) this::createReducingState),
+			Tuple2.of(AggregatingStateDescriptor.class, (StateFactory) this::createAggregatingState),
+			Tuple2.of(FoldingStateDescriptor.class, (StateFactory) this::createFoldingState)
+		).collect(Collectors.toMap(t -> t.f0, t -> t.f1));
+	}
+
+	private interface StateFactory {
+		<N, SV, S extends State, IS extends S> IS create(
+			TypeSerializer<N> namespaceSerializer,
+			StateDescriptor<S, SV> stateDesc) throws Exception;
+	}
+
+	private <N, SV, S extends State, IS extends S> IS createState(
+		TypeSerializer<N> namespaceSerializer,
+		StateDescriptor<S, SV> stateDesc) throws Exception {
+		StateFactory stateFactory = stateFactories.get(stateDesc.getClass());
+		if (stateFactory == null) {
+			String message = String.format("State %s is not supported by %s",
+				stateDesc.getClass(), TtlStateFactory.class);
+			throw new FlinkRuntimeException(message);
+		}
+		return stateFactory.create(namespaceSerializer, stateDesc);
+	}
+
+	@SuppressWarnings("unchecked")
+	private <N, SV, S extends State, IS extends S> IS createValueState(
+		TypeSerializer<N> namespaceSerializer,
+		StateDescriptor<S, SV> stateDesc) throws Exception {
+		SV defVal = stateDesc.getDefaultValue();
+		TtlValue<SV> ttlDefVal = defVal == null ? null : new TtlValue<>(defVal, Long.MAX_VALUE);
+		ValueStateDescriptor<TtlValue<SV>> ttlDescriptor = new ValueStateDescriptor<>(
```

Comment:
Why are we using the deprecated constructor instead of following the suggested replacement of handling `null` values properly ourselves?

Reply:
agree, I think we may actually not expose default value for value state with TTL or new descriptor builders in general